### PR TITLE
maybe fix

### PR DESCRIPTION
**Pull request by yingjieMiao**

### CODE REVIEW

File: kifi-backend/modules/cortex/app/com/keepit/cortex/utils/MatrixUtils.scala [hunk 1]
```diff
@@ -125,8 +125,10 @@ object MatrixUtils {
       val diff = sample(i) - mean(i)
       if (v == 0) s += weights(i) * diff * diff
       else {
-        s += exp(log(weights(i)) + 2 * log(diff) - log(v)) // numerical care
+        val r = weights(i) / v
```
Comment (stkem) [inline kifi-backend/modules/cortex/app/com/keepit/cortex/utils/MatrixUtils.scala:128]: I'm afraid this is also no going to work. You still have the same problem of dividing by a small number.
Suppose the true variance `v` is 10^-9  and the variance calculation has a error of only 10^-7 (which is pretty good!). Then `r` will be off by two orders of magnitude no matter what `weights(i)` is. Assuming a similar (low!) error in the weight, you might easily be looking at an error in the 3-4 orders of magnitude.
In short, small numbers are a problem. This is the reason gaussian elimination uses pivoting. 
There are two well understood techniques for dealing with this. One is regularization and the other is projecting the problem into a subspace where the matrix in question is no longer (numerically) singular.

Comment (yingjieMiao) [inline kifi-backend/modules/cortex/app/com/keepit/cortex/utils/MatrixUtils.scala:128]: the old one gives NaN. I would be surprised if the weights(i) or diff is 
exactly zero. (the weight is actually the mean of several topic 
distributions, which itself is a distribution.

the new code fixed the NaN problem. (and the results are comparable to 
the python results)

There are two sources of error here: measure error and numerical error. 
the measure error might be mitigated if we use a more Bayesian approach.
For the numerical error, I can 'shift' (divide) the variances by the max 
value, and convert things back. Do you think that would make a difference?
